File: internal/fetch/fetch.go
package fetch

import (
	"context"
	"crypto/tls"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"time"

	"github.com/kitstream/initium/internal/safety"
)

type Config struct {
	URL                    string
	OutputPath             string
	Workdir                string
	AuthEnv                string
	InsecureTLS            bool
	FollowRedirects        bool
	AllowCrossSiteRedirect bool
	Timeout                time.Duration
}

func (c Config) Validate() error {
	if c.URL == "" {
		return fmt.Errorf("url is required")
	}
	if c.OutputPath == "" {
		return fmt.Errorf("output is required")
	}
	if c.AllowCrossSiteRedirect && !c.FollowRedirects {
		return fmt.Errorf("--allow-cross-site-redirects requires --follow-redirects")
	}
	return nil
}

func Do(ctx context.Context, cfg Config) error {
	if err := cfg.Validate(); err != nil {
		return err
	}

	outPath, err := safety.ValidateFilePath(cfg.Workdir, cfg.OutputPath)
	if err != nil {
		return fmt.Errorf("invalid output path: %w", err)
	}

	client := buildClient(cfg)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cfg.URL, nil)
	if err != nil {
		return fmt.Errorf("creating request: %w", err)
	}

	if cfg.AuthEnv != "" {
		authVal := os.Getenv(cfg.AuthEnv)
		if authVal == "" {
			return fmt.Errorf("auth env var %q is empty or not set", cfg.AuthEnv)
		}
		req.Header.Set("Authorization", authVal)
	}

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("HTTP request to %s: %w", cfg.URL, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("HTTP %s returned status %d", cfg.URL, resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("reading response body: %w", err)
	}

	if err := os.MkdirAll(filepath.Dir(outPath), 0o755); err != nil {
		return fmt.Errorf("creating output directory: %w", err)
	}

	if err := os.WriteFile(outPath, body, 0o644); err != nil {
		return fmt.Errorf("writing output %s: %w", outPath, err)
	}

	return nil
}

func buildClient(cfg Config) *http.Client {
	transport := &http.Transport{}
	if cfg.InsecureTLS {
		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true} //nolint:gosec // user-opt-in via --insecure-tls
	}

	client := &http.Client{
		Timeout:   cfg.Timeout,
		Transport: transport,
	}

	if !cfg.FollowRedirects {
		client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		}
	} else if !cfg.AllowCrossSiteRedirect {
		client.CheckRedirect = sameSiteRedirectPolicy
	}

	return client
}

func sameSiteRedirectPolicy(req *http.Request, via []*http.Request) error {
	if len(via) >= 10 {
		return fmt.Errorf("too many redirects")
	}
	if len(via) == 0 {
		return nil
	}
	origHost := hostFromURL(via[0].URL)
	newHost := hostFromURL(req.URL)
	if origHost != newHost {
		return fmt.Errorf("cross-site redirect from %s to %s is not allowed; use --allow-cross-site-redirects to permit", origHost, newHost)
	}
	return nil
}

func hostFromURL(u *url.URL) string {
	h := u.Hostname()
	return h
}

File: src/cmd/fetch.rs
use crate::logging::Logger;
use crate::retry;
use crate::safety;
use std::fs;
use std::io::Read;
use std::time::{Duration, Instant};
pub struct Config {
    pub url: String,
    pub output: String,
    pub workdir: String,
    pub auth_env: String,
    pub insecure_tls: bool,
    pub follow_redirects: bool,
    pub allow_cross_site_redirects: bool,
    pub timeout: Duration,
}
impl Config {
    pub fn validate(&self) -> Result<(), String> {
        if self.url.is_empty() { return Err("--url is required".into()); }
        if self.output.is_empty() { return Err("--output is required".into()); }
        if self.allow_cross_site_redirects && !self.follow_redirects {
            return Err("--allow-cross-site-redirects requires --follow-redirects".into());
        }
        Ok(())
    }
}
pub fn run(log: &Logger, cfg: &Config, retry_cfg: &retry::Config) -> Result<(), String> {
    cfg.validate()?;
    let deadline = Instant::now() + cfg.timeout;
    log.info("fetching", &[("url", &cfg.url), ("output", &cfg.output)]);
    let result = retry::do_retry(retry_cfg, Some(deadline), |attempt| {
        log.debug("fetch attempt", &[("attempt", &format!("{}", attempt + 1))]);
        do_fetch(cfg)
    });
    if let Some(e) = result.err {
        log.error("fetch failed", &[("url", &cfg.url), ("error", &e)]);
        return Err(format!("fetch {} failed: {}", cfg.url, e));
    }
    log.info("fetch completed", &[("url", &cfg.url), ("output", &cfg.output), ("attempts", &format!("{}", result.attempt + 1))]);
    Ok(())
}
fn do_fetch(cfg: &Config) -> Result<(), String> {
    let out_path = safety::validate_file_path(&cfg.workdir, &cfg.output)?;
    let agent = if cfg.insecure_tls {
        use std::sync::Arc;
        let crypto_provider = rustls::crypto::ring::default_provider();
        let tls_config = rustls::ClientConfig::builder_with_provider(Arc::new(crypto_provider))
            .with_safe_default_protocol_versions().unwrap()
            .dangerous()
            .with_custom_certificate_verifier(Arc::new(super::wait_for::NoVerifier))
            .with_no_client_auth();
        ureq::AgentBuilder::new()
            .timeout(cfg.timeout)
            .tls_config(Arc::new(tls_config))
            .redirects(if cfg.follow_redirects { 10 } else { 0 })
            .build()
    } else {
        ureq::AgentBuilder::new()
            .timeout(cfg.timeout)
            .redirects(if cfg.follow_redirects { 10 } else { 0 })
            .build()
    };
    let mut req = agent.get(&cfg.url);
    if !cfg.auth_env.is_empty() {
        let auth_val = std::env::var(&cfg.auth_env)
            .map_err(|_| format!("auth env var {:?} is empty or not set", cfg.auth_env))?;
        if auth_val.is_empty() {
            return Err(format!("auth env var {:?} is empty or not set", cfg.auth_env));
        }
        req = req.set("Authorization", &auth_val);
    }
    let resp = req.call().map_err(|e| format!("HTTP request to {}: {}", cfg.url, e))?;
    let status = resp.status();
    if status < 200 || status >= 300 {
        return Err(format!("HTTP {} returned status {}", cfg.url, status));
    }
    let mut body = Vec::new();
    resp.into_reader().read_to_end(&mut body)
        .map_err(|e| format!("reading response body: {}", e))?;
    if let Some(parent) = out_path.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("creating output directory: {}", e))?;
    }
    fs::write(&out_path, &body).map_err(|e| format!("writing output {:?}: {}", out_path, e))?;
    Ok(())
}
